tests/unit/backend/test_base.py

```diff
@@ -0,0 +1,59 @@
+from typing import Iterator, List, Optional
+
+import pytest
+
+from guidellm.backend import Backend, BackendEngine, GenerativeResponse, OpenAIBackend
+from guidellm.core import TextGenerationRequest
+
+
+@Backend.register(backend_type=BackendEngine.TEST)
+class TestBackend(Backend):
+    """
+    The test implementation of a LLM Backend.
+    """
+
+    def __init__(self, target: str, model: str = "test"):
+        self.target: str = target
+        self.model: str = model
+
+    def make_request(
+        self, request: TextGenerationRequest
+    ) -> Iterator[GenerativeResponse]:
+        raise NotImplementedError
+
+    def available_models(self) -> List[str]:
+        raise NotImplementedError
+
+    @property
+    def default_model(self) -> str:
+        raise NotImplementedError
+
+    def model_tokenizer(self, model: str) -> Optional[str]:
+        raise NotImplementedError
+
+
+@pytest.mark.smoke
+def test_backend_registry():
+    """
+    Ensure that all registered classes exist in the registry
+    after they
+    """
+
+    assert Backend._registry == {
+        BackendEngine.TEST: TestBackend,
+        BackendEngine.OPENAI_SERVER: OpenAIBackend,
+    }
```

Comment:
I was wondering about this when I was writing some of my own test. It makes sense that the most effective way to test the registry would be to just directly check the values like this, but I thought because registry is defined as a private variable (_registry vs registry), that it was maybe not the intended way to test this? But I couldn't think of any way to test it without directly accessing the private variable. I think this is more of a philosophical quandary, curious of your thoughts. If this is the best way I'll probably do this as well.

Reply:
@DaltheCow good point here!

I was preparing a huge answer on that but removed the line 😄
We had a discussion on that one with Mark and this `register` decorator will demonstrate its power a bit later.

But here is the topic: if you look for existing tests you will find that they are pretty much straightforward.
I mean the core philosophical question for me here is whether should that be tested or if we can trust Python developers that the Python `__init__` dunder methods will work perfectly in all the cases. And, we should test only the business logic.

So yeah it is more about: are data models ( _aka entities, values objects, aggregates, whatever_ ) should be tested in general.

![Alacritty 2024-07-04](https://github.com/neuralmagic/guidellm/assets/45270625/b2b7d497-9375-460a-b6a4-f220c8f6cf28)


Probably @markurtz may take a hand for both of us here 😁

Reply:
@DaltheCow btw, check `/tests/unit/backend/test_openai.py`

Reply:
Yeah makes sense, I guess to me those tests are pretty low cost to maintain, but also a bit silly to write in the first place. But still I don't mind them.